Signals

tmcgilchrist opened the pull request:
Solution for Stdlib - missing signals #13825

* 1faf314 does the minimal change adding the missing signals
* 46308d8 adds `signal_to_string` to give POSIX names for the signals
* 3a81250 introduces a type alias for signal plus some sanity check tests.

What do you think @nojb @dbuenzli ? This keeps the escape option of supplying your own signal numbers and the type alias documents the intent of the API (without providing much type safety).

tmcgilchrist committed:
Add signal definitions for SIGIO and SIGWINCH.

tmcgilchrist committed:
Add signal_to_string function

tmcgilchrist committed:
Add signal type alias and signal_to_string.

jmid commented on stdlib/sys.mli:
```diff
@@ -240,110 +241,130 @@ type signal_behavior =
-val sigquit : int
-(** Interactive termination *)
+val sigquit : signal
+(** Signaleractive termination *)
```
I think a query-replace `int->signal` went too far here

jmid commented on stdlib/sys.mli:
```diff
@@ -240,110 +241,130 @@ type signal_behavior =
-val sigtstp : int
-(** Interactive stop *)
+val sigtstp : signal
+(** Signaleractive stop *)
```
Same

jmid commented on stdlib/sys.mli:
```diff
@@ -240,110 +241,130 @@ type signal_behavior =
-val sigprof : int
-(** Profiling interrupt *)
+val sigprof : signal
+(** Profiling signalerrupt *)
```
and same

jmid reviewed (commented):
A couple of drive-by comments spotted while skimming the PR

jmid commented on stdlib/sys.mli:
```diff
@@ -240,110 +241,130 @@ type signal_behavior =
-val sigtrap : int
-(** Trace/breakpoint trap
+val sigtrap : signal
+(** Trace/breakposignal trap
```
and here

nojb commented on stdlib/sys.mli:
```diff
@@ -240,110 +241,130 @@ type signal_behavior =
+val signal_to_string : signal -> string
+(** Signal name as a string, based on POSIX names.
+    Raises [Not_found] exception for unrecognised signal.
+    @since 5.4 *)
```
What is the use-case of this function? Debugging? Also, we try to avoid exceptions for new library functions in general.

dbuenzli replied:
> What is the use-case of this function? Debugging?

Reporting signals to end-users, e.g. after a [signaling](https://github.com/ocaml/ocaml/blob/f08e8a1ad48013dbdefc0e5415c2bf48a6881de8/otherlibs/unix/wait.c#L59-L62) `Unix.waitpid`. If you print the bare int, end-users won't be able to make sense of OCaml's renumbering. 

But yes raising `Not_found` is really not a good idea (especially since you don't have a function to check which signals are recognized…). Usually [that's the way](https://github.com/b0-system/b0/blob/93841fa33a9c65f07aef0b5d8982b78d25a2a8aa/src/std/b0__fmt.mli#L209-L212) I do it.

tmcgilchrist replied:
My use case is also for reporting signals to end users, I've written that function across a few different projects and it looks like others have too. I'll remove the exception and just do a default string including the signal number.

dbuenzli replied:
Please document the string format more exactly (see e.g. the doc string I linked to), knowing how the formatting looks is essential when you integrate the string in end user messages.

tmcgilchrist commented on runtime/signals.c:
```diff
@@ -499,12 +499,25 @@ CAMLexport value caml_process_pending_actions_exn(void)
 #ifndef SIGXFSZ
 #define SIGXFSZ -1
 #endif
+#ifndef SIGEMT
+#define SIGEMT -1
+#endif
```
Thanks for the context and history lesson. :-) 
I'll remove it.

tmcgilchrist committed:
Add changelog

dbuenzli commented:
As we discussed in the issue I think it would be nice to add:

```ocaml
val Sys.signal_of_int : int -> signal
(** [signal_of_int n] converts a platform dependent signal number [n] to an OCaml 
    signal number. This is [n] itself if the number is unknown. *)
```

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -227,11 +227,12 @@ external poll_actions : unit -> unit = "%poll"
 
 (** {1 Signal handling} *)
 
+type signal = int
```
A doc string is missing here. This would be a good place to document OCaml's signal renumbering business. Something like:

```ocaml
(** The type for signal numbers. 

    This is either a platform independent negative number for those signals 
    that OCaml recognizes or a positive number for a platform dependent 
    signal number. The function {!signal_of_int} converts known platform dependent 
    number to independent ones.
```

nojb commented on Changes:
```diff
@@ -229,6 +229,11 @@ Working version
 - #13578: On Windows, use the OS CSPRNG to seed the Stdlib.Random generator.
   (Antonin Décimo, review by Miod Vallat, Nicolás Ojeda Bär, and Xavier Leroy)
 
+- #13843: Add signal definitions for SIGIO, SIGWINCH and SIGINFO. Introduces a
```
I think strictly speaking this is a breaking change, so it should be prefixed with an asterisk.
```suggestion
* #13843: Add signal definitions for SIGIO, SIGWINCH. Introduces a
```

nojb reviewed (commented):
@tmcgilchrist The PR looks pretty close to being done. If you can add `signal_of_int` and address the few other comments left, we can move to final review and merge. Thanks!

tmcgilchrist committed:
Export signal conversion functions as primitives in Sys

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -227,11 +227,18 @@ external poll_actions : unit -> unit = "%poll"
+  This is either a platform independent negative number for those signals
+  that OCaml recognizes or a positive number for a platform dependent
+  signal number. The function {!signal_of_int} converts known platform dependent
+  number to independent ones. *)
```
This should also sport an `@since 5.4` annotation.

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -240,110 +247,130 @@ type signal_behavior =
+    signal number. This is [n] itself if the number is unknown.
+    @since 5.4 *)
+
+val signal_to_int : signal -> int
```
This is missing a doc string. For example

```
(** [signal_to_int n] converts a platform independent signal number [n] to a platform
    dependent signal number. This is [n] itself if the number is the number is not 
    a recognized platform independent signal number.
    @since 5.4 *)
```

dbuenzli reviewed (approved):
I have two documentation nits but I think this is good to go.

One thing that is not part of this PR but would be nice to add is to update `Unix`'s `.mli` with the `Sys.signal` type where appropriate. This should be [here](https://github.com/ocaml/ocaml/blob/c67b1830597e84b5c48f08ff340d592f7cdb95a0/otherlibs/unix/unix.mli#L1150-L1214) and [there](https://github.com/ocaml/ocaml/blob/c67b1830597e84b5c48f08ff340d592f7cdb95a0/otherlibs/unix/unix.mli#L193-L198). 

Also [perhaps](https://github.com/ocaml/ocaml/pull/13843/checks#step:10:568) this needs a bootstrap ?

tmcgilchrist committed:
Export signal conversion functions as primitives.

dbuenzli commented on Changes:
```diff
@@ -229,6 +229,12 @@ Working version
 - #13578: On Windows, use the OS CSPRNG to seed the Stdlib.Random generator.
   (Antonin Décimo, review by Miod Vallat, Nicolás Ojeda Bär, and Xavier Leroy)
 
+* #13843: Add signal definitions for SIGIO, SIGWINCH and SIGINFO. Introduces a
```
`SIGINFO` is not added.

dbuenzli commented on stdlib/sys.ml.in:
```diff
@@ -109,6 +111,49 @@ let sigtrap = -25
+  | s' when s' == sigwinch -> "SIGWINCH"
+  | s' -> "SIG(" ^ (string_of_int s') ^ ")"
+
+external rev_convert_signal_number: int -> int = "caml_sys_rev_convert_signal_number"
```
This is exceeding 80 columns which is making check-typo complain (and the CI fail).

tmcgilchrist committed:
Update documentation in Unix module

tmcgilchrist commented on tools/sync_stdlib_docs:
```diff
@@ -25,7 +25,7 @@ fi
 #Removes a label, i.e a space, a variable name, followed by a colon followed by
 #an alphabetic character or ( or '. This should avoid altering the contents of
 #comments.
-LABREGEX="s/ [a-z_]+:([a-z\('])/ \1/g"
```
This change is required to match `signal:Sys.signal` in `unixLabels.mli` previously there were no Module.type references.

tmcgilchrist commented:
This is ready for review again @dbuenzli @nojb 

I've changed the Unix documentation to refer to `Sys.signal` and tidied up the other comments.

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -227,11 +227,20 @@ external poll_actions : unit -> unit = "%poll"
+  This is either a platform independent negative number for those signals
+  that OCaml recognizes or a positive number for a platform dependent
+  signal number. The function {!signal_of_int} converts known platform dependent
+  numbers to independent ones, and {!signal_to_int} does the reverse converting
```
You could have stopped with a dot after the `reverse`.

dbuenzli commented on otherlibs/unix/unix.mli:
```diff
@@ -190,15 +190,15 @@ type process_status =
 (** The termination status of a process.  See module {!Sys} for the
     definitions of the standard signal numbers.  Note that they are
-    not the numbers used by the OS.
+    not the same numbers used by the OS.
```
This wording reads strange to me but you are the native english speaking person.

tmcgilchrist replied:
I think highlighting that the numbers aren't the same as those used by the OS is important. Maybe "Note they are different to the numbers used by the OS." is a shorter way to say the same thing.

I won't insist if this is clear enough to others.

dbuenzli replied:
The `Sys.signal` types mentioned in the variant should now be clickable so I think would rather just change:

> See module {!Sys} for the definitions of the standard signal numbers
> Note that they are not the numbers used by the OS.

To

> Note that signal numbers reported by the OS are converted to platform independent {!Sys.signal} values.

Where all the details are written in the doc string of `Sys.signal`.

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -240,110 +249,134 @@ type signal_behavior =
+val signal_of_int : int -> signal
+(** [signal_of_int n] converts a platform dependent signal number [n] to
+    an OCaml signal number.
+    This is [n] itself if the number is unknown.
```
Since POSIX interface generally use `int` for signal numbers on the C side and this is what is used by the conversion functions this is not exactly true on 64-bit platforms. The integers are going to be truncated you should get e.g. `Sys.signal_of_int max_int = Int32.(to_int max_int)`. Not sure it's worth mentioning though.

tmcgilchrist replied:
I agree. Using large int values for signal numbers is unusual.

dbuenzli commented on stdlib/sys.mli:
```diff
@@ -240,110 +249,134 @@ type signal_behavior =
+val signal_to_int : signal -> int
+(** [signal_to_int n] converts an OCaml signal number [n] to
+    a platform dependent signal number.
+    This is [n] itself if the number is unknown.
```
The same comment applies here.

dbuenzli reviewed (approved):
Appart from a few doc nits. This looks good to me. Thanks for having added the changes to `Unix`.

tmcgilchrist commented:
@nojb Are you ok with the changes here?

nojb commented on otherlibs/unix/unix.mli:
```diff
@@ -190,15 +190,14 @@ type process_status =
-(** The termination status of a process.  See module {!Sys} for the
-    definitions of the standard signal numbers.  Note that they are
-    not the numbers used by the OS.
+(** The termination status of a process.  See module {!Sys.signal} for the
```
```suggestion
(** The termination status of a process.  See {!Sys.signal} for the
```

nojb commented on otherlibs/unix/unixLabels.mli:
```diff
@@ -190,15 +190,14 @@ type process_status = Unix.process_status =
-(** The termination status of a process.  See module {!Sys} for the
-    definitions of the standard signal numbers.  Note that they are
-    not the numbers used by the OS.
+(** The termination status of a process.  See module {!Sys.signal} for the
```
```suggestion
(** The termination status of a process.  See {!Sys.signal} for the
```

nojb commented on stdlib/sys.ml.in:
```diff
@@ -109,6 +111,51 @@ let sigtrap = -25
+  | s' when s' == sigxfsz -> "SIGXFSZ"
+  | s' when s' == sigio -> "SIGIO"
+  | s' when s' == sigwinch -> "SIGWINCH"
+  | s' -> "SIG(" ^ (string_of_int s') ^ ")"
```
```suggestion
  | s' -> "SIG(" ^ string_of_int s' ^ ")"
```

nojb reviewed (approved):
LGTM

- A second approval from a core dev is needed before merging; and
- The bootstrap needs to be redone by a core dev.

Thanks!

MisterDA commented on testsuite/tests/lib-sys/signal.ml:
```diff
@@ -0,0 +1,85 @@
+(* TEST
+ include unix;
+ libunix;
+ native;
```
```suggestion
 include unix;
 hasunix;
 not-windows;
 native;
```

MisterDA reviewed (commented):
Would it make sense to mark `Sys.signal` as `private`? As we have `signal_of_int` this could avoid mistakes.

nojb commented:
> Would it make sense to mark `Sys.signal` as `private`? As we have `signal_of_int` this could avoid mistakes.

Unfortunately, this would break user code, which goes against the strong backwards-compatibility guarantees the standard library usually abides by.

dra27 commented on stdlib/sys.ml.in:
```diff
@@ -109,6 +111,51 @@ let sigtrap = -25
+
+let signal_of_int i = rev_convert_signal_number i
+
+let signal_to_int i = convert_signal_number i
```
Just a simplicity nit:

```suggestion
external signal_of_int: int -> int =
  "caml_sys_rev_convert_signal_number"
external signal_to_int: int -> int =
  "caml_sys_convert_signal_number"
```

dra27 commented on stdlib/sys.ml.in:
```diff
@@ -109,6 +111,51 @@ let sigtrap = -25
+
+let signal_to_string s =
+  match s with
+  | s' when s' == sigabrt -> "SIGABRT"
```
Physical equality on ints? FWIW, given that they generate the same code, what's wrong with:
```ocaml
if s = sigabrt then
  "SIGABRT"
else if s = (* ... *)
```
?

dra27 commented on stdlib/sys.mli:
```diff
@@ -227,11 +227,19 @@ external poll_actions : unit -> unit = "%poll"
+  This is either a platform independent negative number for those signals
+  that OCaml recognizes or a positive number for a platform dependent
+  signal number. The function {!signal_of_int} converts known platform dependent
+  numbers to independent ones, and {!signal_to_int} does the reverse.
```
Possible alternative:
```suggestion
  Negative numbers are used by OCaml to a provide a platform-independent
  number for signals recognised by OCaml. Positive numbers are always the
  platform-dependent value for a given signal.
  The function {!signal_of_int} converts known platform dependent
  numbers to independent ones, and {!signal_to_int} does the reverse.
```

dra27 commented on stdlib/sys.mli:
```diff
@@ -227,11 +227,19 @@ external poll_actions : unit -> unit = "%poll"
   | Signal_ignore
-  | Signal_handle of (int -> unit)   (** *)
+  | Signal_handle of (signal -> unit)   (** *)
 (** What to do when receiving a signal:
```
Worth clarifying here that the negative values are always passed - i.e. that for known signals, the value provided to the function is always one of the constants below?

dra27 commented on stdlib/sys.mli:
```diff
@@ -240,116 +248,139 @@ type signal_behavior =
    first argument is the signal number.  Return the behavior
    previously associated with the signal. If the signal number is
    invalid (or not available on your system), an [Invalid_argument]
    exception is raised. *)
```
I think it's worth being explicit either here, or in the general comments, that you can specify a positive value for the signal number here, but that OCaml will always use the negative value for signals it recognised - and that we have `signal_of_int` for this.

dra27 commented on stdlib/sys.mli:
```diff
@@ -240,116 +248,139 @@ type signal_behavior =
+
+val signal_to_string : signal -> string
+(** [signal_to_string] formats an OCaml [signal] as a C POSIX
+    {{:http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/signal.h.html}
```
Issue 8:
```suggestion
    {{:https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/signal.h.html}
```

dra27 commented on stdlib/sys.mli:
```diff
@@ -240,116 +248,139 @@ type signal_behavior =
+val signal_of_int : int -> signal
+(** [signal_of_int n] converts a platform dependent signal number [n] to
+    an OCaml signal number.
+    This is [n] itself if the number is unknown.
```
"unknown" feels nebulous:
```suggestion
    This is [n] itself if OCaml does not have a platform-independent signal number for [n].
```

dra27 commented on stdlib/sys.mli:
```diff
@@ -240,116 +248,139 @@ type signal_behavior =
+    @since 5.4 *)
+
+val signal_of_int : int -> signal
+(** [signal_of_int n] converts a platform dependent signal number [n] to
```
```suggestion
(** [signal_of_int n] converts a platform-dependent signal number [n] to
```
(and below - cf. https://en.wikipedia.org/wiki/Wikipedia:How_to_make_dashes#Platform-dependent_methods 😁)

dra27 commented on stdlib/sys.mli:
```diff
@@ -240,116 +248,139 @@ type signal_behavior =
+(** Window size change
+    @since 5.4 *)
+
+val signal_to_string : signal -> string
```
I'm not sure that these functions should ignore unknown _negative_ numbers - the C conversion functions do that for memory safety, knowing that they shouldn't see them. `signal_to_string` and `signal_to_int` should call `invalid_arg` if presented with an unrecognised _negative_ number (that's a programming error somewhere).

Shouldn't `signal_of_int` call `invalid_arg` for _all_ negative numbers - similarly a programming error?

tmcgilchrist replied:
I think that is sensible, I've made the changes in ab2569a2ee.

dra27 reviewed (commented):
I don't know if it's a subsequent change in the PR, but the bootstrap doesn't look necessary to me (it's fully passed without it on my fork) - can it be removed?

Various comments, but virtually all documentation nits. The only salient point is that I don't think we should silently ignore unrecognised/invalid _negative_ signal numbers ... unless there's a compelling reason I'm missing?

tmcgilchrist committed:
Check int and signal argument ranges for conversion functions

Signal_to_string and signal_to_int call invalid_arg if presented with
an unrecognised negative number. signal_of_int calls invalid_arg for
all negative numbers.

dra27 commented on Changes:
```diff
@@ -252,6 +252,13 @@ Working version
+  type alias for signal int, signal_to_string to convert OCaml signal numbers
+  to their POSIX equivalent names, and signal_of_int/signal_to_int for
+  converting between OCaml and platform signal numbers. (Reported in #13825)
+  (Tim McGilchrist, review by Daniel Bünzli, Nicolás Ojeda Bär, Miod Vallat
+   and Jan Midtgaard)
```
```suggestion
  (Tim McGilchrist, review by David Allsopp, Nicolás Ojeda Bär, Daniel Bünzli
   Jan Midtgaard and Miod Vallat)
```

dra27 reviewed (approved):
LGTM, thank you!

MisterDA commented:
Did we miss that POSIX 2024 added [`sig2str`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/sig2str.html)? Support before this standard varies from system to system. Maybe we should have used that, with a fallback to our own code.

dra27 commented:
It's not an appropriate function to use here - we expect `Sys.signal_to_string` to work regardless of the underlying library (even for signals the platform doesn't implement). In particular, the `Sys.signal_to_string Sys.sig*` should behave consistently on Windows (and some of the signals are of course emulated/implemented).
